nemo/collections/common/prompts/canary2.py

```diff
@@ -0,0 +1,206 @@
+        slot_values = map_manifest_values_to_special_tokens(slot_values)
+        return super().encode_turn(
+            prompt_template=prompt_template, expected_slots=expected_slots, slot_values=slot_values
+        )
+
+
+def map_manifest_values_to_special_tokens(slot_values: dict[str, str]) -> dict[str, str]:
+    slot_values = slot_values.copy()
+
+    any_special_token_present = False
+
+    lang_dict_compat = {"en": "en-US", "es": "es-ES", "fr": "fr-FR", "de": "de-DE"}
+    for k in ("source_lang", "target_lang"):
+        if k in slot_values and not ((v := slot_values[k]).startswith("<|") and v.endswith("|>")):
+            val = slot_values[k]
+            val = lang_dict_compat.get(val, val)
+            slot_values[k] = "<|" + val + "|>"
+            any_special_token_present = True
+
+    # Handle boolean slots
+    for k in ("pnc", "itn", "timestamp", "diarize"):
+        true_token = f"<|{k}|>"
+        false_token = f"<|no{k}|>"
+        if k in slot_values and slot_values[k] not in (true_token, false_token):
+            slot_values[k] = true_token if slot_values[k] in ("yes", "1", "True", "true", k) else false_token
+            any_special_token_present = True
+
+    # Auto-inject which tokenizer to look up in CanaryTokenizer if not provided,
+    # and slots for this turn correspond to user prompt.
+    if any_special_token_present and PromptFormatter.PROMPT_LANGUAGE_SLOT not in slot_values:
+        slot_values[PromptFormatter.PROMPT_LANGUAGE_SLOT] = CANARY_SPECIAL_TOKENIZER
+    else:
+        if (l := slot_values.get(PromptFormatter.PROMPT_LANGUAGE_SLOT)) is not None:
+            slot_values[PromptFormatter.PROMPT_LANGUAGE_SLOT] = lang_dict_compat.get(l, l)
+
+    return slot_values
+
+
+@registered_prompt_format_fn
+def canary2(
+    cuts: CutSet, tokenizer: TokenizerSpec
+) -> tuple[list[torch.Tensor], list[torch.Tensor], list[torch.Tensor]]:
+    """
+    Prepend and append control tokens to the token sequence as per Canary 2.0 format.
+
+    The prompt format syntax is defined in :class:`Canary2PromptFormatter`
+    """
+    formatter = Canary2PromptFormatter(tokenizer)
+
+    prompts_with_answers, prompts, answers = [], [], []
+    for cut in cuts:
+        if isinstance(cut, MixedCut):
+            cut = cut._first_non_padding_cut
+        if not isinstance(cut, MonoCut):
+            raise TypeError(
+                f"Expected input audio to have a single channel (required MonoCut/MixedCut, but we received: {cut=})"
+            )
+
+        # first, validate the utterance
+        expected_slots = {"source_lang", "target_lang", "pnc"}
+        missing_keys = expected_slots - set(cut.custom)
+        if missing_keys:
+            raise RuntimeError(
+                f"We found cut with ID {cut.id} that is missing the following keys: {missing_keys}"
+                f"Please ensure that every utterance in the input manifests contains these keys."
+            )
+
+        optional_slots = {
```

Comment:
Is there a check to maintain consistency across slots or are you allowing variable slot sizes? (e.g. sample 1 has `itn` sample 2 does not have that slot.)

Reply:
we are combining different datasets annotated for different things, so the flexibility is required atleast the level of each manifest if not per sample.

Reply:
Hmm, most of these are binary prompts. What about just having default values for the prompts instead? Else I believe y'all are adding the extra difficulty of the model needing to learn variable length prompt input.

Reply:
There is no variable length (except if you add decoder context). We indeed have default values for slots here.